### PR TITLE
refactor lib npm.js

### PR DESCRIPTION
**Pull request by isaacs**
Cleaning up `lib/npm.js`, `lib/help.js`, `bin/npm-cli.js`, and creating 100% test coverage along the way.

Come and get it!  Packed full of refactoring and shiny new unit tests.

### CODE REVIEW

**Comment by ruyadorno** [comment]
@isaacs given the new default for `npm ls` I think it would be nice to also set `all: true` for either `npm ll` or `npm la` in order to make them more useful 😊

**Comment by isaacs** [comment]
This PR fixes #1506 and #1533.

File: lib/utils/proc-log-listener.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+module.exports = () => {
+  process.on('log', (level, ...args) => {
+    try {
+      log[level](...args)
+    } catch (ex) {
+      try {
+        // if it crashed once, it might again!
+        log.verbose(`attempt to log ${inspect([level, ...args])} crashed`, ex)
+      } catch (ex2) {
+        console.error(`attempt to log ${inspect([level, ...args])} crashed`, ex)
```
Comment (ruyadorno) [inline lib/utils/proc-log-listener.js:12]:
```suggestion
        console.error(`attempt to log ${inspect([level, ...args])} crashed`, ex2)
```

**Review by ruyadorno — APPROVED** [review]
LGTM 👍 :shipit:

**Pull request closed without merging**